src/permutat.cc

```diff
@@ -2649,150 +2649,109 @@ static Obj FuncMappingPermListList(Obj self, Obj src, Obj dst)
-            if (dt >= nn)
-              for ( i = 0; i < nn; i++)
-                ptR[i] = (UInt2) ptT[ptR[i]];
-            else
-              for ( i = 0; i < nn; i++)
-                ptR[i] = (UInt2)IMAGE(ptR[i], ptT, dt);
-          }
-          im = (Int)ptR[bpt];
-        } else {
-          UInt4 *ptR = ADDR_PERM4(result);
-          if (IS_PERM2(t)) {
-            const UInt2 *ptT = CONST_ADDR_PERM2(t);
-            UInt dt = DEG_PERM2(t);
-            if (dt >= nn)
-              for ( i = 0; i < nn; i++)
-                ptR[i] = (UInt4)ptT[ptR[i]];
-            else
-              for ( i = 0; i < nn; i++)
-                ptR[i] = (UInt4)IMAGE(ptR[i], ptT, dt);
-          }
-          else {
-            const UInt4 *ptT = CONST_ADDR_PERM4(t);
-            UInt dt = DEG_PERM4(t);
-            if (dt >= nn)
-              for ( i = 0; i < nn; i++)
-                ptR[i] = ptT[ptR[i]];
-            else
-              for ( i = 0; i < nn; i++)
-                ptR[i] = IMAGE(ptR[i], ptT, dt);
-          }
-          im = (Int)ptR[bpt];
+
+    while (IsbPRec(stb, RN_stabilizer)) {
+        Obj trans = ElmPRec(stb, RN_transversal);
+        Obj orb = ElmPRec(stb, RN_orbit);
+        Int bpt = INT_INTOBJ(ELM_LIST(orb, 1)) - 1;
+
+        const Res * ptrResult = CONST_ADDR_PERM<Res>(result);
+        UInt        degResult = DEG_PERM<Res>(result);
+        Int         im = (Int)(IMAGE(bpt, ptrResult, degResult));
+        Obj         t = ELM0_LIST(trans, im + 1);
+
+        if (!t)
+            break;
+
+        while (bpt != im) {
+
+            ptR = ADDR_PERM<Res>(result);
+            if (IS_PERM2(t)) {
+                const UInt2 * ptT = CONST_ADDR_PERM2(t);
+                UInt          dt = DEG_PERM2(t);
+                if (dt >= nn)
+                    for (i = 0; i < nn; i++)
+                        ptR[i] = ptT[ptR[i]];
+                else
+                    for (i = 0; i < nn; i++)
+                        ptR[i] = IMAGE(ptR[i], ptT, dt);
+            }
+            else {
+                const UInt4 * ptT = CONST_ADDR_PERM4(t);
+                UInt          dt = DEG_PERM4(t);
+                if (dt >= nn)
+                    for (i = 0; i < nn; i++)
+                        ptR[i] = ptT[ptR[i]];
+                else
+                    for (i = 0; i < nn; i++)
+                        ptR[i] = IMAGE(ptR[i], ptT, dt);
+            }
+            im = (Int)ptR[bpt];
+            t = ELM_PLIST(trans, im + 1);
         }
-        t = ELM_PLIST(trans,im+1);
-      }
+        stb = ElmPRec(stb, RN_stabilizer);
+    }
+    /* so we're done sifting, and now we just have to clean up result */
+    return result;
+}
+
+static Obj FuncSCR_SIFT_HELPER(Obj self, Obj stb, Obj g, Obj n)
+{
+    if (!IS_PREC(stb))
+        RequireArgument("SCR_SIFT_HELPER", stb, "must be a plain record");
+    RequirePermutation("SCR_SIFT_HELPER", g);
+    UInt nn = GetSmallInt("SCR_SIFT_HELPER", n);
+
+    if (!RN_stabilizer)
```

Comment:
Seems a bit weird to put these in here, instead of SCR_SIFT_HELPER, which actually needs/uses them?

Reply:
This is shared code, so not putting it in there reduces the size of the generated code. Also, IMO it feels a bit less cluttered... In fact, strictly speaking, I'd prefer to init `RN_stabilizer` in `PostRestore`... I'll adjust the PR.